### PR TITLE
Enable per-sample `fill_value` argument in Erase operator 

### PR DESCRIPTION
**Pull request by jantonguirao**
Signed-off-by: Joaquin Anton <janton@nvidia.com>

## Category:

**New feature**

## Description:

- It enables a new feature in Erase operator. `fill_value` argument can be fed with a tensor argument and not only a constant value
- Adjusted EraseGpu kernel to include the fill values in the sample descriptors
- Modified the Erase operator to support the per-sample argument.
- Added Kernel tests and operator tests
- Added several Run overloads to EraseGpu kernel to support invocation with constant, CPU tensor list view, and GPU tensor list view, for convenience.
- Requested by https://github.com/NVIDIA/DALI/issues/3881

### Affected modules and functionalities:

Erase operator

### Key points relevant for the review:

Changes in the kernel and operator? Can we simplify something?

### Tests:

dali/kernels/erase/erase_gpu_test.cu (see testcases with PER_SAMPLE_CPU and PER_SAMPLE_GPU)
dali/test/python/operator/test_erase.py

- [ ] Existing tests apply
- [x] New tests added
  - [x] Python tests
  - [x] GTests
  - [ ] Benchmark
  - [ ] Other
- [ ] N/A

### Documentation
- [ ] Existing documentation applies
- [ ] Documentation updated
  - [ ] Docstring
  - [ ] Doxygen
  - [ ] RST
  - [ ] Jupyter
  - [ ] Other
- [x] N/A

#### Requirements
- [ ] Implements new requirements
- [x] Affects existing requirements
- [ ] N/A
Extended the requirement
**REQ IDs**: ERASE.07

**JIRA TASK**: DALI-2969

### CODE REVIEW

File: dali/kernels/erase/erase_gpu.h [hunk 1]
```diff
@@ -507,20 +480,73 @@ struct EraseGpu {
...
     erase_gpu_impl<channel_dim><<<grid_dim, block_dim, 0, stream>>>(
-        sample_desc_gpu, region_dim, fill_values_span);
+        sample_desc_gpu, region_dim);
+  }
+
+  void Run(KernelContext &ctx,
+           OutListGPU<T, ndim> &out,
+           const InListGPU<T, ndim> &in,
+           const InListGPU<ibox<ndim>, 1> &erased_regions,
+           span<const float> fill_values) {
```
Comment (jantonguirao) [inline dali/kernels/erase/erase_gpu.h:498]: overload for constant fill values

File: dali/kernels/erase/erase_gpu.h [hunk 2]
```diff
@@ -507,20 +480,73 @@ struct EraseGpu {
...
+    TensorListView<StorageCPU, const float, 1> fill_values_tlv;
+    fill_values_tlv.resize(in.num_samples());
+    for (int i = 0; i < in.num_samples(); i++) {
+      fill_values_tlv.data[i] = fill_values.data();
+      fill_values_tlv.shape.tensor_shape_span(i)[0] = fill_values.size();
+    }
+    Run(ctx, out, in, erased_regions, fill_values_tlv);
+  }
+
+  template <typename T2>
```
Comment (jantonguirao) [inline dali/kernels/erase/erase_gpu.h:508]: overload for CPU fill values tensor list (as opposed to GPU)

**Review by prak-nv — APPROVED** [review]

**Review by awolant — APPROVED** [review]

**Comment by jantonguirao** [comment]
!build

**Comment by jantonguirao** [comment]
!build

**Pull request merged**